onnxruntime/core/providers/cuda/cuda_graph.cc

```diff
@@ -0,0 +1,82 @@
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License.
+
+#include "core/providers/cuda/cuda_graph.h"
+
+#include "core/providers/cuda/cuda_common.h"
+#include <cuda_runtime_api.h>
+#include <driver_types.h>
+
+
+namespace onnxruntime {
+
+CUDAGraph::CUDAGraph(cudaStream_t stream) : capture_stream_(stream) {
+#if (defined(CUDA_VERSION) && CUDA_VERSION < 10000)
+  ORT_THROW("CUDA graphs can only be used in Onnxruntime built with CUDA >= 10.0");
+#endif
+  CUDA_CALL_THROW(cudaStreamCreateWithFlags(&replay_stream_, cudaStreamNonBlocking));
+}
+
+void CUDAGraph::CaptureBegin() {
+#if defined(CUDA_VERSION) && CUDA_VERSION >= 10000
+  ORT_ENFORCE(!has_graph_exec_,
+              "This cuda graph has already captured a graph. "
+              "Create a new instance to capture a new graph.");
+
+  CUDA_CALL_THROW(cudaDeviceSynchronize());
+  CUDA_CALL_THROW(cudaStreamBeginCapture(capture_stream_, cudaStreamCaptureModeGlobal));
+#else
+  ORT_THROW("CUDA graphs can only be used in Onnxruntime built with CUDA >= 10.0");
+#endif
+}
+
+void CUDAGraph::CaptureEnd() {
+#if defined(CUDA_VERSION) && CUDA_VERSION >= 10000
+  CUDA_CALL_THROW(cudaStreamEndCapture(capture_stream_, &graph_));
+  if (graph_ == NULL) {
+    ORT_THROW("CUDAGraph::CaptureEnd: graph_ is NULL");
+  }
+
+  has_graph_ = true;
+  CUDA_CALL_THROW(cudaGraphInstantiate(&graph_exec_, graph_, NULL, NULL, 0));
+  has_graph_exec_ = true;
+  CUDA_CALL_THROW(cudaGraphDestroy(graph_));
+  has_graph_ = false;
+#else
+  ORT_THROW("CUDA graphs can only be used in Onnxruntime built with CUDA >= 10.0");
+#endif
+}
+
+Status CUDAGraph::Replay() {
+  std::lock_guard<OrtMutex> lock(lock_);
```

Comment:
Is this lock necessary now that there is a CUDAGraph instance per-thread in the CUDA EP ? I think it can be removed with a comment saying that Replay() is not thread-safe ?

Reply:
No, it is not needed. I will remove it and add a comment there.